### PR TITLE
Rebase Ethermint w/ Optimint v0.3.0 on tharsis/ethermint v0.16.0

### PR DESCRIPTION
**Pull request by jbowen93**
Rebase Optimint v0.3.0 on tharsis/ethermint v0.16.0 #61

Failing tests are gosec and markdown lint which aren't an issue.

resolves https://github.com/celestiaorg/ethermint/issues/49

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -18,29 +19,30 @@ require (
...
-	google.golang.org/protobuf v1.27.1
+	golang.org/x/crypto v0.0.0-20220411220226-7b82a4e95df4 // indirect
+	google.golang.org/genproto v0.0.0-20220407144326-9054f6ed7bac
+	google.golang.org/grpc v1.45.0
+	google.golang.org/protobuf v1.28.0
 	gopkg.in/yaml.v2 v2.4.0
 )
 
-require github.com/celestiaorg/optimint v0.2.0
+require github.com/celestiaorg/optimint v0.3.0
```
Comment (tzdybal) [inline go.mod:45]: You should use [v0.3.1](https://github.com/celestiaorg/optimint/releases/tag/v0.3.1)

**Review by adlerjohn — COMMENTED** [review]
Can you get CI to not fail?

**Comment by jbowen93** [comment]
> Can you get CI to not fail?

Not really. 
- Markdown lint is because of auto formatting of changelogs on tharsis side
- The protobuf breakage isn't even in this PR (there's no diff on proto/ethermint/evm/v1/evm.proto)
- The gosec file inclusion errors are only because those files got renamed along the way so they went through the CI again

File: .github/workflows/docker-build-test.yml [hunk 2]
```diff
@@ -64,6 +64,7 @@ jobs:
       with:
         repository: celestiaorg/ephemeral-cluster
         path: ephemeral-cluster
+        ref: jbowen93/mess
```
Comment (adlerjohn) [inline .github/workflows/docker-build-test.yml:67]: Is this expected?

File: CHANGELOG.md [hunk 3]
```diff
@@ -36,6 +36,137 @@ Ref: https://keepachangelog.com/en/1.0.0/
 
 # Changelog
 
+<<<<<<< HEAD
```
Comment (adlerjohn) [inline CHANGELOG.md:39]: This looks like the result of a merge conflict.

File: CHANGELOG.md [hunk 4]
```diff
@@ -61,6 +197,7 @@ Ref: https://keepachangelog.com/en/1.0.0/
 
 ### API Breaking
 
+>>>>>>> v0.16.0
```
Comment (adlerjohn) [inline CHANGELOG.md:200]: Ditto

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -59,17 +58,21 @@ jobs:
     runs-on: ubuntu-latest
     timeout-minutes: 10
     steps:
-      - uses: actions/setup-go@v2.1.4
+      - uses: actions/setup-go@v3
         with:
           go-version: 1.17
+<<<<<<< HEAD
```
Comment (adlerjohn) [inline .github/workflows/test.yml:64]: Ditto

File: CHANGELOG.md [hunk 6]
```diff
@@ -36,6 +36,137 @@ Ref: https://keepachangelog.com/en/1.0.0/
...
+* (feemarket) [tharsis#1104](https://github.com/tharsis/ethermint/pull/1104) Enforce a minimum gas price for Cosmos and EVM transactions through the `MinGasPrice` parameter.
+* (rpc) [tharsis#1081](https://github.com/tharsis/ethermint/pull/1081) Deduplicate some json-rpc logic codes, cleanup several dead functions.
+* (ante) [tharsis#1062](https://github.com/tharsis/ethermint/pull/1062) Emit event of eth tx hash in ante handler to support query failed transactions.
+* (analytics) [tharsis#1106](https://github.com/tharsis/ethermint/pull/1106) Update telemetry to Ethermint modules.
+* (rpc) [tharsis#1108](https://github.com/tharsis/ethermint/pull/1108) Update GetGasPrice RPC endpoint with global `MinGasPrice`
+>>>>>>> v0.16.0
+
+* (ante) [tharsis#964](https://github.com/tharsis/ethermint/pull/964) add NewInfiniteGasMeterWithLimit for storing the user provided gas limit. Fixes block's consumed gas calculation in the block creation phase.
+
+<<<<<<< HEAD
```
Comment (adlerjohn) [inline CHANGELOG.md:61]: Ditto

File: .github/workflows/docker-build-test.yml (hunk 2 above)
Comment (jbowen93) [reply .github/workflows/docker-build-test.yml:67]: Yes 😓

**Review by adlerjohn — COMMENTED** [review]
The CI jobs `Tests / install-tparse` and `Tests / test-unit-cover` should be fixed.

**Review by adlerjohn — COMMENTED** [review]
Is it expected for the docker builds and unit test workflows to fail?

**Comment by jbowen93** [comment]
> Is it expected for the docker builds and unit test workflows to fail?

Docker builds failed because I tried to use the "fast build" method that gets around using qemu which causes ~32 minute build times. It used to work, now it doesn't. 

Opening a [separate issue](https://github.com/celestiaorg/ethermint/issues/68) for this and going back to the slow qemu method.

**Comment by adlerjohn** [comment]
How about the `test-unit-cover`?

**Comment by jbowen93** [comment]
> How about the `test-unit-cover`?

That should be passing, digging into why it's not.

**Comment by jbowen93** [comment]
The new `TestPendingTransactionFilter` doesn't compile for some reason. I'm going to consider that part of the [already existing issue](https://github.com/celestiaorg/ethermint/issues/12) and add a TODO

**Comment by jbowen93** [comment]
@adlerjohn Apologies for all of the delays and back and forth here but I believe this PR is in a mergeable state now. 

The remaining oddity is [here](https://github.com/celestiaorg/ethermint/pull/61/files?show-deleted-files=true&show-viewed-files=true&file-filters%5B%5D=#diff-6c98f963e459783869600db95ea3119087f32552eaac89f0e074e9bff22c439eR67) where we test using the `jbowen93/mess` branch of the `ephemeral-cluster`. That's because I haven't gotten around to cleaning up [this PR](https://github.com/celestiaorg/ephemeral-cluster/pull/25) to be merge ready.

**Review by adlerjohn — APPROVED** [review]
utACK

**Pull request merged**